README.md
[![Published on NPM](https://img.shields.io/npm/v/@advanced-rest-client/date-time.svg)](https://www.npmjs.com/package/@advanced-rest-client/date-time)

[![Build Status](https://travis-ci.org/advanced-rest-client/date-time.svg?branch=stage)](https://travis-ci.org/advanced-rest-client/date-time)

[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/advanced-rest-client/date-time)

## &lt;date-time&gt;

An element to render data and/or time formatted for user locale.

Note, this component is pure web component. Polymer is used for demo page, tests, and documentation.


```html
<date-time date="2010-12-10T11:50:45Z" year="numeric" month="narrow" day="numeric"></date-time>
```

### API components

This components is a part of [API components ecosystem](https://elements.advancedrestclient.com/)

## Usage

### Installation
```
npm install --save @advanced-rest-client/date-time
```

### In an html file

```html
<html>
  <head>
    <script type="module">
      import '@advanced-rest-client/date-time/date-time.js';
    </script>
  </head>
  <body>
    <date-time></date-time>
  </body>
</html>
```

### In a Polymer 3 element

```js
import {PolymerElement, html} from '@polymer/polymer';
import '@advanced-rest-client/date-time/date-time.js';

class SampleElement extends PolymerElement {
  static get template() {
    return html`
    <date-time></date-time>
    `;
  }

  _authChanged(e) {
    console.log(e.detail);
  }
}
customElements.define('sample-element', SampleElement);
```

### Installation

```sh
git clone https://github.com/advanced-rest-client/date-time
cd api-url-editor
npm install
npm install -g polymer-cli
```

### Running the demo locally

```sh
polymer serve --npm
open http://127.0.0.1:<port>/demo/
```

### Running the tests
```sh
polymer test --npm
```


CHANGELOG.md
<a name="2.0.4"></a>
## [2.0.4](https://github.com/advanced-rest-client/date-time/compare/2.0.1...2.0.4) (2018-10-29)




<a name="2.0.3"></a>
## [2.0.3](https://github.com/advanced-rest-client/date-time/compare/2.0.1...2.0.3) (2018-04-16)




<a name="2.0.2"></a>
## [2.0.2](https://github.com/advanced-rest-client/date-time/compare/2.0.1...2.0.2) (2018-04-15)




<a name="2.0.2"></a>
## [2.0.2](https://github.com/advanced-rest-client/date-time/compare/2.0.1...2.0.2) (2018-03-11)




<a name="0.1.2"></a>
## [0.1.2](https://github.com/advanced-rest-client/date-time/compare/0.1.1...v0.1.2) (2017-01-02)




<a name="0.1.1"></a>
## 0.1.1 (2017-01-02)




<a name="0.0.2"></a>
## 0.0.2 (2017-01-02)




# [3.0.0](https://github.com/advanced-rest-client/date-time/compare/2.0.1...3.0.0) (2019-06-11)




## [3.0.1](https://github.com/advanced-rest-client/date-time/compare/2.0.1...3.0.1) (2019-07-19)


### build

* bumping version ([d9767fa0cc09f35629b0bfcf2922ae7c54ba3dae](https://github.com/advanced-rest-client/date-time/commit/d9767fa0cc09f35629b0bfcf2922ae7c54ba3dae))

### chore

* updating project's metadata ([d7e17cdf7127d14902252bf15803181fda08048b](https://github.com/advanced-rest-client/date-time/commit/d7e17cdf7127d14902252bf15803181fda08048b))

### refactor

* adding <time> element for accessibility support ([65fd679c97046c5a39374ae51a66606dcff8c0df](https://github.com/advanced-rest-client/date-time/commit/65fd679c97046c5a39374ae51a66606dcff8c0df))

### test

* fixing test name in sauce labs ([3a315bf84673d60856dc2663ddd76e9fd18bd08e](https://github.com/advanced-rest-client/date-time/commit/3a315bf84673d60856dc2663ddd76e9fd18bd08e))



## [3.0.2](https://github.com/advanced-rest-client/date-time/compare/2.0.1...3.0.2) (2019-08-31)





test/date-time.test.js
import { fixture, assert, nextFrame } from '@open-wc/testing';
import * as sinon from 'sinon/pkg/sinon-esm.js';
import '../date-time.js';

describe('<date-time>', () => {
  async function basicFixture() {
    return (await fixture(`
      <date-time locales="en-US" time-zone="UTC" date="2010-12-10T11:05:45.000Z"></date-time>`));
  }

  async function longWeekdayFixture() {
    return (await fixture(`
      <date-time locales="en-US" time-zone="UTC" date="2010-12-10T11:05:45.000Z" weekday="long"></date-time>`));
  }

  async function longMonthFixture() {
    return (await fixture(`
      <date-time locales="en-US" time-zone="UTC" date="2010-12-10T11:05:45.000Z" month="long"></date-time>`));
  }

  async function numericYearFixture() {
    return (await fixture(`
      <date-time locales="en-US" time-zone="UTC" date="2010-12-10T11:05:45.000Z" year="numeric"></date-time>`));
  }

  async function numericDayFixture() {
    return (await fixture(`
      <date-time locales="en-US" time-zone="UTC" date="2010-12-10T11:05:45.000Z" day="numeric"></date-time>`));
  }

  async function numericHourFixture() {
    return (await fixture(`
      <date-time locales="en-US" time-zone="UTC" date="2010-12-10T11:05:45.000Z" hour="numeric"></date-time>`));
  }

  async function numericMinuteFixture() {
    return (await fixture(`
      <date-time locales="en-US" time-zone="UTC" date="2010-12-10T11:05:45.000Z" minute="numeric"></date-time>`));
  }

  async function numericSecondFixture() {
    return (await fixture(`
      <date-time
        locales="en-US"
        time-zone="UTC"
        date="2010-12-10T11:05:45.000Z"
        second="numeric"></date-time>`));
  }

  async function emptyFixture() {
    return (await fixture(`<date-time></date-time>`));
  }

  async function deutscheFixture() {
    return (await fixture(`
      <date-time locales="de-DE" time-zone="UTC" date="2010-12-10T11:05:45.000Z"></date-time>`));
  }

  async function itempropFixture() {
    return (await fixture(`
      <date-time itemprop="title" date="2010-12-10T11:05:45.000Z"></date-time>`));
  }

  const EDGE_IS_STILL_SO_BAD = /\u200E/g;
  function normalizeString(str) {
    return str.replace(EDGE_IS_STILL_SO_BAD, '');
  }
  const hasSupport = typeof Intl !== 'undefined';
  describe('Basics', () => {
    it('Renders date string value', async () => {
      // Can't check date format at this point since it may vary depending
      // on the locale settings.
      const element = await basicFixture();
      await nextFrame();
      const txt = element._getTimeNode().innerHTML;
      assert.typeOf(txt, 'string');
    });

    it('Should compute ISO time', async () => {
      const element = await basicFixture();
      await nextFrame();
      assert.equal(element._getTimeNode().getAttribute('datetime'), '2010-12-10T11:05:45.000Z');
    });

    it('Should set weekday', async () => {
      if (hasSupport) {
        const element = await longWeekdayFixture();
        await nextFrame();
        const txt = element._getTimeNode().innerHTML;
        assert.equal(normalizeString(txt), 'Friday');
      }
    });

    it('Sets year', async () => {
      if (hasSupport) {
        const element = await numericYearFixture();
        const txt = element._getTimeNode().innerHTML;
        assert.equal(normalizeString(txt), '2010');
      }
    });

    it('Should set month', async () => {
      if (hasSupport) {
        const element = await longMonthFixture();
        const txt = element._getTimeNode().innerHTML;
        assert.equal(normalizeString(txt), 'December');
      }
    });

    it('Should set day', async () => {
      if (hasSupport) {
        const element = await numericDayFixture();
        const txt = element._getTimeNode().innerHTML;
        assert.equal(normalizeString(txt), '10');
      }
    });

    it('Should set hour', async () => {
      if (hasSupport) {
        const element = await numericHourFixture();
        const txt = element._getTimeNode().innerHTML;
        assert.include(normalizeString(txt), '11');
      }
    });

    it('Sets minute', async () => {
      if (hasSupport) {
        const element = await numericMinuteFixture();
        const txt = element._getTimeNode().innerHTML;
        assert.equal(normalizeString(txt), '5');
      }
    });

    it('Sets minute', async () => {
      if (hasSupport) {
        const element = await numericSecondFixture();
        const txt = element._getTimeNode().innerHTML;
        assert.equal(normalizeString(txt), '45');
      }
    });
  });

  describe('Attributes settings', () => {
    let element;
    beforeEach(async () => {
      element = await emptyFixture();
    });

    [
      ['locales'], ['date'], ['year'], ['month'], ['day'], ['hour'], ['minute'], ['second'],
      ['weekday', undefined, 'short'], ['time-zone-name', 'timeZoneName', 'short'],
      ['era', undefined, 'narrow'], ['time-zone', 'timeZone', 'UTC'], ['hour12', undefined, true]
    ].forEach((item) => {
      it(`Calls _updateLabel() when "${item[0]}" attribute changes`, () => {
        const value = item[2] || 'numeric';
        const spy = sinon.spy(element, '_updateLabel');
        element.setAttribute(item[0], value);
        assert.isTrue(spy.called);
      });

      it(`Updates propety value when "${item[0]}" attribute changes`, () => {
        const prop = item[1] || item[0];
        const value = item[2] || 'numeric';
        element.setAttribute(item[0], value);
        if (prop === 'hour12') {
          assert.isTrue(element[prop]);
        } else {
          assert.equal(element[prop], value);
        }
      });
    });
  });

  describe('Properies changed', () => {
    let element;
    beforeEach(async () => {
      element = await emptyFixture();
    });

    [
      ['locales'], ['date'], ['year'], ['month'], ['day'], ['hour'], ['minute'], ['second'],
      ['weekday', undefined, 'short'], ['time-zone-name', 'timeZoneName', 'short'],
      ['era', undefined, 'narrow'], ['time-zone', 'timeZone', 'UTC'], ['hour12', undefined, true]
    ].forEach((item) => {
      it(`Calls _updateLabel() when "${item[0]}" property change`, () => {
        const value = item[2] || 'numeric';
        const prop = item[1] || item[0];
        const spy = sinon.spy(element, '_updateLabel');
        element[prop] = value;
        assert.isTrue(spy.called);
      });

      it(`Updates attribute value when "${item[0]}" property changes`, () => {
        const attr = item[0];
        const prop = item[1] || attr;
        const value = item[2] || 'numeric';
        element[prop] = value;
        assert.isTrue(element.hasAttribute(attr), 'Has corresponding attribute');
        if (prop !== 'hour12') {
          assert.equal(element.getAttribute(attr), value);
        }
      });
    });
  });

  describe('_getIntlOptions()', () => {
    let element;
    beforeEach(async () => {
      element = await emptyFixture();
    });

    [
      ['year'], ['month'], ['day'], ['hour'], ['minute'], ['second'],
      ['weekday', undefined, 'short'], ['time-zone-name', 'timeZoneName', 'short'],
      ['era', undefined, 'narrow'], ['time-zone', 'timeZone', 'UTC'], ['hour12', undefined, true]
    ].forEach((item) => {
      it(`Adds "${item[0]}" property`, () => {
        const attr = item[0];
        const prop = item[1] || attr;
        const value = item[2] || 'numeric';
        element[prop] = value;
        const result = element._getIntlOptions();
        if (prop === 'hour12') {
          assert.isTrue(result[prop]);
        } else {
          assert.equal(result[prop], value);
        }
      });
    });
  });

  describe('_updateLabel()', () => {
    let element;

    it('Does nothing when not in the DOM', async () => {
      element = await basicFixture();
      const parent = element.parentElement;
      parent.removeChild(element);
      const time = element._getTimeNode();
      element.shadowRoot.removeChild(time);
      element.day = 'numeric';
      const node = element.shadowRoot.querySelector('time');
      assert.notOk(node);
    });

    it('Sets "datetime" attribiute on <time>', async () => {
      element = await basicFixture();
      const time = element._getTimeNode();
      assert.equal(time.getAttribute('datetime'), '2010-12-10T11:05:45.000Z');
    });

    it('Sets text content on <time>', async () => {
      element = await basicFixture();
      const time = element._getTimeNode();
      assert.equal(time.innerText.trim().toLowerCase(), '12/10/2010');
    });

    it('Uses different locale', async () => {
      element = await deutscheFixture();
      const time = element._getTimeNode();
      assert.equal(time.innerText.trim().toLowerCase(), '10.12.2010');
    });
  });

  describe('itemprop attribute', () => {
    let element;

    it('Coppies itemprop attribute to <time> element', async () => {
      element = await itempropFixture();
      assert.equal(element._getTimeNode().getAttribute('itemprop'), 'title');
    });

    it('Removes itemprop attribute from this element', async () => {
      element = await itempropFixture();
      assert.equal(element.getAttribute('itemprop'), null);
    });

    it('Ignores the change when attribute is already set', async () => {
      element = await itempropFixture();
      const spy = sinon.spy(element, '_getTimeNode');
      element.itemprop = 'title';
      // Getter uses `_getTimeNode` to read attribute.
      // This should happen only once in theis case.
      assert.isTrue(spy.calledOnce);
    });
  });
});


date-time.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/tools/tree/master/packages/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   date-time.html
 */

declare namespace UiElements {

  /**
   * An element to display formatted date and time.
   *
   * The `date` propery accepts Date object, Number as a timestamp or string
   * that will be parsed to the Date object.
   *
   * This element uses the `Intl` interface which is available in IE 11+ browsers.
   *
   * To format the date use [Intl.DateTimeFormat]
   * (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat)
   * inteface options.
   *
   * The default value for each date-time component property is undefined,
   * but if all component properties are undefined, then year, month, and day
   * are assumed to be "numeric" (per spec).
   *
   * ### Example
   * ```html
   * <date-time date="2010-12-10T11:50:45Z" year="numeric" month="narrow" day="numeric"></date-time>
   * ```
   *
   * The element provides accessibility by using the `time` element and setting
   * the `datetime` attribute on it.
   *
   * ### Styling
   *
   * `<date-time>` provides the following custom properties and mixins for styling:
   *
   * Custom property | Description | Default
   * ----------------|-------------|----------
   * `--date-time` | Mixin applied to the element | `{}`
   */
  class DateTime extends HTMLElement {

    /**
     * A string with a BCP 47 language tag, or an array of such strings.
     * For the general form and interpretation of the locales argument,
     * see the Intl page.
     * The following Unicode extension keys are allowed:
     * - nu - Numbering system. Possible values include: "arab", "arabext",
     * "bali", "beng", "deva", "fullwide", "gujr", "guru", "hanidec", "khmr",
     * "knda", "laoo", "latn", "limb", "mlym", "mong", "mymr", "orya",
     * "tamldec", "telu", "thai", "tibt".
     * - ca - Calendar. Possible values include: "buddhist", "chinese",
     * "coptic", "ethioaa", "ethiopic", "gregory", "hebrew", "indian",
     * "islamic", "islamicc", "iso8601", "japanese", "persian", "roc".
     */
    locales: any;

    /**
     * The representation of the year.
     * Possible values are "numeric", "2-digit".
     *    
     */
    year: any;

    /**
     * The representation of the month.
     * Possible values are "numeric", "2-digit", "narrow", "short", "long".
     *    
     */
    month: any;

    /**
     * The representation of the day.
     * Possible values are "numeric", "2-digit".
     *    
     */
    day: any;

    /**
     * The representation of the hour.
     * Possible values are "numeric", "2-digit".
     *    
     */
    hour: any;

    /**
     * The representation of the minute.
     * Possible values are "numeric", "2-digit".
     *    
     */
    minute: any;

    /**
     * The representation of the second.
     * Possible values are "numeric", "2-digit".
     *    
     */
    second: any;

    /**
     * The representation of the weekday.
     * Possible values are "narrow", "short", "long".
     *    
     */
    weekday: any;

    /**
     * The representation of the time zone name.
     *
     * Possible values are "short", "long".
     *    
     */
    timeZoneName: any;

    /**
     * The time zone to use. The only value implementations must recognize
     * is "UTC"; the default is the runtime's default time zone.
     * Implementations may also recognize the time zone names of the IANA
     * time zone database, such as "Asia/Shanghai", "Asia/Kolkata",
     * "America/New_York".
     *    
     */
    timeZone: any;

    /**
     * The representation of the era.
     *
     * Possible values are "narrow", "short", "long".
     *    
     */
    era: any;

    /**
     * Whether to use 12-hour time (as opposed to 24-hour time).
     * Possible values are `true` and `false`; the default is locale
     * dependent.
     */
    hour12: any;

    /**
     * A date object to render.
     * It can be a `Date` object, number representing a timestamp
     * or valid date string. The argument is parsed by `Date` constructor
     * to produce the value.
     */
    date: any;
    connectedCallback(): void;
    disconnectedCallback(): void;
    attributeChangedCallback(): void;

    /**
     * Parses input `date` to a Date object.
     *
     * @param date A date to parse
     */
    _getParsableDate(date: String|Number|Date|null): Date|null;
    _getIntlOptions(): any;
    _updateLabel(): void;
    _setIso(v: any): void;
  }
}

interface HTMLElementTagNameMap {
  "date-time": UiElements.DateTime;
}


karma.sl.config.js
/* eslint-disable import/no-extraneous-dependencies */
const merge = require('deepmerge');
const slSettings = require('@advanced-rest-client/testing-karma-sl/sl-settings.js');
const createBaseConfig = require('./karma.conf.js');

module.exports = (config) => {
  const cnf = slSettings();
  cnf.sauceLabs.testName = 'date-type';
  cnf.browsers = [
    'SL_Chrome',
    'SL_Chrome-1',
    'SL_Firefox',
    'SL_Firefox-1',
    'SL_Safari',
    // 'SL_EDGE'
  ];
  if (process.env.TRAVIS) {
    const buildLabel = 'TRAVIS #' + process.env.TRAVIS_BUILD_NUMBER + ' (' + process.env.TRAVIS_BUILD_ID + ')';

    cnf.browserStack = {
      build: buildLabel,
      tunnelIdentifier: process.env.TRAVIS_JOB_NUMBER
    };

    cnf.sauceLabs.build = buildLabel;
    cnf.sauceLabs.tunnelIdentifier = process.env.TRAVIS_JOB_NUMBER;
  }

  config.set(merge(createBaseConfig(config), cnf));

  return config;
};


date-time.js
/**
@license
Copyright 2016 The Advanced REST client authors <arc@mulesoft.com>
Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
*/
/**
 * An element to display formatted date and time.
 *
 * The `date` propery accepts Date object, Number as a timestamp or string
 * that will be parsed to the Date object.
 *
 * This element uses the `Intl` interface which is available in IE 11+ browsers.
 *
 * To format the date use [Intl.DateTimeFormat]
 * (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat)
 * inteface options.
 *
 * The default value for each date-time component property is undefined,
 * but if all component properties are undefined, then year, month, and day
 * are assumed to be "numeric" (per spec).
 *
 * ### Example
 *
 * ```html
 * <date-time date="2010-12-10T11:50:45Z" year="numeric" month="narrow" day="numeric"></date-time>
 * ```
 *
 * The element provides accessibility by using the `time` element and setting
 * the `datetime` attribute on it.
 *
 * ### Styling
 *
 * `<date-time>` provides the following custom properties and mixins for styling:
 *
 * Custom property | Description | Default
 * ----------------|-------------|----------
 * `--date-time` | Mixin applied to the element | `{}`
 *
 *
 * @customElement
 * @demo demo/index.html
 * @memberof UiElements
 */
class DateTime extends HTMLElement {
  static get observedAttributes() {
    return [
      'locales', 'date', 'year', 'month', 'day', 'hour', 'minute', 'second',
      'weekday', 'time-zone-name', 'era', 'time-zone', 'hour12', 'itemprop'
    ];
  }

  constructor() {
    super();
    this.attachShadow({ mode: 'open' });
    this._observer = new MutationObserver(() => {
      this.setAttribute('aria-label', this.shadowRoot.textContent);
    });
  }

  connectedCallback() {
    this._observer.observe(this.shadowRoot, {
      childList: true,
      characterData: true,
      subtree: true
    });
    this._updateLabel();
  }

  disconnectedCallback() {
    this._observer.disconnect();
  }
  /**
   * A string with a BCP 47 language tag, or an array of such strings.
   * For the general form and interpretation of the locales argument,
   * see the Intl page.
   * The following Unicode extension keys are allowed:
   * - nu - Numbering system. Possible values include: "arab", "arabext",
   * "bali", "beng", "deva", "fullwide", "gujr", "guru", "hanidec", "khmr",
   * "knda", "laoo", "latn", "limb", "mlym", "mong", "mymr", "orya",
   * "tamldec", "telu", "thai", "tibt".
   * - ca - Calendar. Possible values include: "buddhist", "chinese",
   * "coptic", "ethioaa", "ethiopic", "gregory", "hebrew", "indian",
   * "islamic", "islamicc", "iso8601", "japanese", "persian", "roc".
   *
   * @type {String}
   */
  get locales() {
    return this.getAttribute('locales');
  }
  /**
   * A string with a BCP 47 language tag, or an array of such strings.
   * For the general form and interpretation of the locales argument,
   * see the Intl page.
   * The following Unicode extension keys are allowed:
   * - nu - Numbering system. Possible values include: "arab", "arabext",
   * "bali", "beng", "deva", "fullwide", "gujr", "guru", "hanidec", "khmr",
   * "knda", "laoo", "latn", "limb", "mlym", "mong", "mymr", "orya",
   * "tamldec", "telu", "thai", "tibt".
   * - ca - Calendar. Possible values include: "buddhist", "chinese",
   * "coptic", "ethioaa", "ethiopic", "gregory", "hebrew", "indian",
   * "islamic", "islamicc", "iso8601", "japanese", "persian", "roc".
   *
   * @param {String} v
   * @type {String}
   */
  set locales(v) {
    this.setAttribute('locales', v);
  }
  /**
   * The representation of the year.
   * Possible values are "numeric", "2-digit".
   */
  get year() {
    return this.getAttribute('year');
  }
  /**
   * The representation of the year.
   * @param {String} v Possible values are "numeric", "2-digit".
   */
  set year(v) {
    this.setAttribute('year', v);
  }
  /**
   * The representation of the month.
   * Possible values are "numeric", "2-digit", "narrow", "short", "long".
   */
  get month() {
    return this.getAttribute('month');
  }
  /**
   * The representation of the month.
   * @param {String} v Possible values are "numeric", "2-digit", "narrow", "short", "long".
   */
  set month(v) {
    this.setAttribute('month', v);
  }
  /**
   * The representation of the day.
   * Possible values are "numeric", "2-digit".
   */
  get day() {
    return this.getAttribute('day');
  }
  /**
   * The representation of the day.
   * @param {String} v Possible values are "numeric", "2-digit".
   */
  set day(v) {
    this.setAttribute('day', v);
  }
  /**
   * The representation of the hour.
   * Possible values are "numeric", "2-digit".
   */
  get hour() {
    return this.getAttribute('hour');
  }
  /**
   * The representation of the hour.
   * @param {String} v Possible values are "numeric", "2-digit".
   */
  set hour(v) {
    this.setAttribute('hour', v);
  }
  /**
   * The representation of the minute.
   * Possible values are "numeric", "2-digit".
   */
  get minute() {
    return this.getAttribute('minute');
  }
  /**
   * The representation of the minute.
   * @param {String} v Possible values are "numeric", "2-digit".
   */
  set minute(v) {
    this.setAttribute('minute', v);
  }
  /**
   * The representation of the second.
   * Possible values are "numeric", "2-digit".
   */
  get second() {
    return this.getAttribute('second');
  }
  /**
   * The representation of the second.
   * @param {String} v Possible values are "numeric", "2-digit".
   */
  set second(v) {
    this.setAttribute('second', v);
  }
  /**
   * The representation of the weekday.
   * Possible values are "narrow", "short", "long".
   */
  get weekday() {
    return this.getAttribute('weekday');
  }
  /**
   * The representation of the weekday.
   * @param {String} v Possible values are "narrow", "short", "long".
   */
  set weekday(v) {
    this.setAttribute('weekday', v);
  }
  /**
   * The representation of the time zone name.
   *
   * Possible values are "short", "long".
   */
  get timeZoneName() {
    return this.getAttribute('time-zone-name');
  }
  /**
   * The representation of the time zone name.
   *
   * @param {String} v Possible values are "short", "long".
   */
  set timeZoneName(v) {
    this.setAttribute('time-zone-name', v);
  }
  /**
   * The time zone to use. The only value implementations must recognize
   * is "UTC"; the default is the runtime's default time zone.
   * Implementations may also recognize the time zone names of the IANA
   * time zone database, such as "Asia/Shanghai", "Asia/Kolkata",
   * "America/New_York".
   */
  get timeZone() {
    return this.getAttribute('time-zone');
  }
  /**
   * The time zone to use. The only value implementations must recognize
   * is "UTC"; the default is the runtime's default time zone.
   * Implementations may also recognize the time zone names of the IANA
   * time zone database, such as "Asia/Shanghai", "Asia/Kolkata",
   * "America/New_York".
   * @param {String} v
   */
  set timeZone(v) {
    this.setAttribute('time-zone', v);
  }
  /**
   * The representation of the era.
   *
   * Possible values are "narrow", "short", "long".
   */
  get era() {
    return this.getAttribute('era');
  }
  /**
   * The representation of the era.
   *
   * @param {String} v Possible values are "narrow", "short", "long".
   */
  set era(v) {
    this.setAttribute('era', v);
  }
  /**
   * Whether to use 12-hour time (as opposed to 24-hour time).
   * Possible values are `true` and `false`; the default is locale
   * dependent.
   *
   * @type {Boolean}
   */
  get hour12() {
    if (!this.hasAttribute('hour12') && !this.__hour12set) {
      return;
    }
    return this.hasAttribute('hour12');
  }
  /**
   * Whether to use 12-hour time (as opposed to 24-hour time).
   * Possible values are `true` and `false`; the default is locale
   * dependent.
   *
   * @param {Boolean} v
   */
  set hour12(v) {
    this.__hour12set = true;
    if (v) {
      this.setAttribute('hour12', '');
    } else {
      this.removeAttribute('hour12');
    }
  }
  /**
   * A date object to render.
   * It can be a `Date` object, number representing a timestamp
   * or valid date string. The argument is parsed by `Date` constructor
   * to produce the value.
   *
   * @type {Date|String|number}
   */
  get date() {
    if (this.__date) {
      return this.__date;
    }
    return this.getAttribute('date');
  }
  /**
   * A date object to render.
   * It can be a `Date` object, number representing a timestamp
   * or valid date string. The argument is parsed by `Date` constructor
   * to produce the value.
   *
   * @param {Date|String|number} v The date to render
   */
  set date(v) {
    this.__date = v;
    if (typeof v === 'string') {
      this.setAttribute('date', v);
    } else {
      this._updateLabel();
    }
  }

  get itemprop() {
    return this._getTimeNode().getAttribute('itemprop');
  }

  set itemprop(value) {
    const old = this.itemprop;
    if (old === value) {
      return;
    }
    if (old && value === null) {
      // This setter moves attribute from this element to "<time>" elsement.
      // When the attribute is removed from this then it becomes null.
      return;
    }
    const node = this._getTimeNode();
    if (value) {
      node.setAttribute('itemprop', value);
      this.removeAttribute('itemprop');
    } else {
      node.removeAttribute('itemprop');
    }
  }

  attributeChangedCallback(name, oldValue, newValue) {
    if (name === 'itemprop') {
      this[name] = newValue;
      return;
    }
    this._updateLabel();
  }
  /**
   * Parses input `date` to a Date object.
   * @param {String|Number|Date} date A date to parse
   * @return {Date}
   */
  _getParsableDate(date) {
    if (!date) {
      date = new Date();
    } else if (typeof date === 'string') {
      try {
        date = new Date(date);
        const _test = date.getDate();
        if (_test !== _test) {
          date = new Date();
        }
      } catch (e) {
        date = new Date();
      }
    } else if (!isNaN(date)) {
      date = new Date(date);
    } else if (!(date instanceof Date)) {
      date = new Date();
    }
    return date;
  }

  _getIntlOptions() {
    const options = {};
    if (this.year) {
      options.year = this.year;
    }
    if (this.month) {
      options.month = this.month;
    }
    if (this.day) {
      options.day = this.day;
    }
    if (this.hour) {
      options.hour = this.hour;
    }
    if (this.minute) {
      options.minute = this.minute;
    }
    if (this.second) {
      options.second = this.second;
    }
    if (this.weekday) {
      options.weekday = this.weekday;
    }
    if (this.era) {
      options.era = this.era;
    }
    if (this.timeZoneName) {
      options.timeZoneName = this.timeZoneName;
    }
    if (this.timeZone) {
      options.timeZone = this.timeZone;
    }
    if (this.hour12 !== undefined) {
      options.hour12 = this.hour12;
    }
    return options;
  }
  /**
   * @return {Element} A reference to a `<time>` element that is in the shadow DOM of this element.
   */
  _getTimeNode() {
    let node = this.shadowRoot.querySelector('time');
    if (!node) {
      node = document.createElement('time');
      this.shadowRoot.appendChild(node);
    }
    return node;
  }

  _updateLabel() {
    if (!this.parentElement) {
      return;
    }
    const date = this._getParsableDate(this.date);
    const node = this._getTimeNode();
    node.setAttribute('datetime', date.toISOString());
    /* istanbul ignore if */
    if (typeof Intl === 'undefined') {
      node.innerText = date.toString();
      return;
    }
    let locales;
    if (this.locales) {
      locales = this.locales;
    }
    const options = this._getIntlOptions();
    const value = new Intl.DateTimeFormat(locales, options).format(date);
    node.innerText = value;
  }
}
window.customElements.define('date-time', DateTime);
